File: sdk/typescript/src/embedding.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import { EmbeddingClient, EmbeddingNotConfiguredError } from "./embedding";

const realFetch = globalThis.fetch;

function resetEnv() {
  delete process.env.KS_GATEWAY_URL;
  delete process.env.KS_RELAY_TOKEN;
  delete process.env.KS_EMBEDDING_MODEL;
  delete process.env.KS_EMBEDDING_DIM;
}

afterEach(() => {
  globalThis.fetch = realFetch;
  resetEnv();
});

describe("EmbeddingClient", () => {
  it("POSTs to /v1/mcp/relay/embeddings", async () => {
    process.env.KS_GATEWAY_URL = "http://gw";
    process.env.KS_RELAY_TOKEN = "tk";
    process.env.KS_EMBEDDING_MODEL = "bge-m3";
    const captured: any = {};
    globalThis.fetch = vi.fn(async (url: any, init: any) => {
      captured.url = String(url);
      captured.init = init;
      return new Response(JSON.stringify({
        object: "list",
        model: "bge-m3",
        data: [{ object: "embedding", index: 0, embedding: [0.1, 0.2] }],
        usage: { prompt_tokens: 2, total_tokens: 2 },
      }), { status: 200, headers: { "Content-Type": "application/json" } });
    }) as any;

    const got = await new EmbeddingClient().embed("hello");
    expect(captured.url).toBe("http://gw/v1/mcp/relay/embeddings");
    expect(captured.init.headers.Authorization).toBe("Bearer tk");
    expect(got.dense).toEqual([0.1, 0.2]);
    expect(got.tokens).toBe(2);
  });

  it("throws when token is missing", async () => {
    process.env.KS_EMBEDDING_MODEL = "bge-m3";
    await expect(new EmbeddingClient().embed("x")).rejects.toBeInstanceOf(EmbeddingNotConfiguredError);
  });
});

File: sdk/go/ksapp/embedding_test.go
package ksapp

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestEmbeddingClientEmbedMany(t *testing.T) {
	var capturedPath, capturedAuth string
	var captured map[string]any
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		capturedPath = r.URL.Path
		capturedAuth = r.Header.Get("Authorization")
		if err := json.NewDecoder(r.Body).Decode(&captured); err != nil {
			t.Fatal(err)
		}
		_, _ = w.Write([]byte(`{"object":"list","model":"bge-m3","data":[{"object":"embedding","index":0,"embedding":[0.1,0.2],"sparse_embedding":{"5":0.7}}],"usage":{"prompt_tokens":2,"total_tokens":2}}`))
	}))
	defer srv.Close()

	c := &EmbeddingClient{gatewayURL: srv.URL, relayToken: "tk", model: "bge-m3", dim: 1024, httpClient: srv.Client()}
	got, err := c.EmbedMany(context.Background(), []string{"hello"})
	if err != nil {
		t.Fatal(err)
	}
	if capturedPath != "/v1/mcp/relay/embeddings" {
		t.Fatalf("path=%s", capturedPath)
	}
	if capturedAuth != "Bearer tk" {
		t.Fatalf("auth=%s", capturedAuth)
	}
	if captured["model"] != "bge-m3" || captured["encoding_format"] != "dense+sparse" {
		t.Fatalf("body=%#v", captured)
	}
	if len(got) != 1 || len(got[0].Dense) != 2 || got[0].Tokens != 2 {
		t.Fatalf("unexpected result: %#v", got)
	}
	if got[0].Sparse["5"] != 0.7 {
		t.Fatalf("sparse not parsed: %#v", got[0].Sparse)
	}
}

func TestEmbeddingClientNoToken(t *testing.T) {
	c := &EmbeddingClient{gatewayURL: "http://x", model: "bge-m3"}
	if _, err := c.Embed(context.Background(), "x"); err == nil {
		t.Fatal("expected error")
	}
}
